grammar.js

```diff
@@ -32,237 +43,304 @@ const stringEncoding = /(u|U)8/;
+
+    [$.parameter, $.declaration_expression],
 
-    [$.assignment_expression, $._expression],
+    [$._reserved_identifier, $.modifier],
+    [$._reserved_identifier, $.scoped_type],
+    [$._reserved_identifier, $.implicit_type],
+    [$._reserved_identifier, $.parameter],
+    [$._reserved_identifier, $.parameter, $.scoped_type],
+    [$._reserved_identifier, $.from_clause],
+    [$._reserved_identifier, $.implicit_type, $.var_pattern],
+    [$._reserved_identifier, $.type_parameter_constraint],
+
+    [$.type_pattern, $.recursive_pattern],
+  ],
+
+  externals: $ => [
+    $._optional_semi,
+    $.interpolation_dollar,
+    $.interpolation_open_brace,
+    $.interpolation_close_brace,
+    $.interpolation_starting_at,
+    $.interpolation_string_content,
+  ],
 
+  extras: $ => [
+    /[\s\u00A0\uFEFF\u3000]+/,
+    $.comment,
+    $.preproc_call,
   ],
 
   inline: $ => [
-    $._identifier_or_global,
+    $._namespace_member_declaration,
+    $._object_creation_type,
+    $._nullable_base_type,
+  ],
+
+  precedences: $ => [
+    [$._anonymous_object_member_declarator, $._simple_name],
+
+    [$.block, $.initializer_expression],
+  ],
+
+  supertypes: $ => [
+    $.declaration,
+    $.expression,
+    $.non_lvalue_expression,
+    $.lvalue_expression,
+    $.literal,
+    $.statement,
+    $.type,
+    $.type_declaration,
+    $.pattern,
   ],
 
   word: $ => $._identifier_token,
 
   rules: {
     compilation_unit: $ => seq(
+      optional($.shebang_directive),
       repeat($.extern_alias_directive),
       repeat($.using_directive),
-      repeat($.global_attribute_list),
+      repeat(choice($.global_attribute, alias($.preproc_if_of_global_attribute, $.preproc_if))),
       choice(
-        seq(repeat($.global_statement), repeat($._namespace_member_declaration)),
-        $.file_scoped_namespace_declaration
-      )
-    ),
-
-    global_statement: $ => $._statement,
-
-    _declaration: $ => choice(
-      $.class_declaration,
-      $.constructor_declaration,
-      $.conversion_operator_declaration,
-      $.delegate_declaration,
-      $.destructor_declaration,
-      $.enum_declaration,
-      $.event_declaration,
-      $.event_field_declaration,
-      $.field_declaration,
-      $.indexer_declaration,
-      $.interface_declaration,
-      $.method_declaration,
-      $.namespace_declaration,
-      $.operator_declaration,
-      $.property_declaration,
-      $.record_declaration,
-      $.record_struct_declaration,
-      $.struct_declaration,
-      $.using_directive,
-    ),
-
-    _namespace_member_declaration: $ => choice(
-      $.namespace_declaration,
-      $._type_declaration
-    ),
-
-    _type_declaration: $ => choice(
-      $.class_declaration,
-      $.struct_declaration,
-      $.interface_declaration,
-      $.enum_declaration,
-      $.delegate_declaration,
-      $.record_declaration,
-      $.record_struct_declaration
+        seq(
+          repeat(alias($.statement, $.global_statement)),
```

Comment:
This change means that `node-types.json` does not provide any information about what `global_statement` can be. Can we revert it back to `repeat($.global_statement)` with `global_statement: $ => $.statement`?

Reply:
I could, but what's the benefit? Curious

Reply:
We have [a generic Treesitter-based CodeQL extractor framework](https://github.com/github/codeql/blob/main/shared/tree-sitter-extractor/src/generator/mod.rs), which, given a Treesitter grammar (`node-types.json`), produces a matching CodeQL database schema. For example, given `node-types.json` [for Ruby](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/src/node-types.json), we produce [a matching DB schema](https://github.com/github/codeql/blob/main/ruby/ql/lib/ruby.dbscheme#L105). When I point it at the `node-types.json` file in this PR, it currently fails because it doesn't know the type of `global_statement`, and hence it is not able to generate a type union that represent all the possible nodes under a `compilation_unit`. In comparison, the [Ruby `program` rule](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/grammar.js#L108) gives rise to [this](https://github.com/github/codeql/blob/main/ruby/ql/lib/ruby.dbscheme#L997) union type.

Reply:
I guess one way we can work around this is basically to say then that a `compilation_unit` can contain _any_ type of AST node as a child.

Reply:
Gotcha, thanks. I'm okay with adding the rule back, though it would be a smidge better without it (one less reduction)

Reply:
In that case I think it would be nice to add it back in, thanks.